### PR TITLE
[Storage-OCI] Support OCI Object Storage [Done]

### PR DESCRIPTION
**Pull request by HysunHe**
Implement #2072 

This PR is to add Oracle Object Storage support. 

Currently, it does not support data move between s3 bucket/gcs bucket and OCI bucket. It will throw NotImplementedError in the data_transfer interfaces.

Tested (run the relevant ones):

- [v] Code formatting: `bash format.sh`
- [v] Any manual or new tests for this PR (please specify below)
    - _get_bucket
    - _create_oci_bucket
    - sky storage ls
    - sky storage delete "skybucket*"
    - sky launch with file_mounts (sky launch -c test013 store.yaml)
    - MOUNT mode
    - COPY mode
    
   -- Test re-uses a predefined bucket 
    file_mounts:
      /datasets-oci: oci://skybucket17

   -- Test multi-source and COPY mode.
    file_mounts:
      /datasets-storage-copy:
        name: skybucket18
        source: ['/home/hysunhe/python/dev/mylog', '/home/hysunhe/rust/dev/modsample']
        store: oci
        persistent: True
        mode: COPY

   -- Test single source and MOUNT mode.
    file_mounts:
      /dataset-storage-mount:
        name: skybucket19
        source: /home/hysunhe/python/dev/mylog
        mode: MOUNT

   -- Test MOUNT-only
    file_mounts:
      /huge-dataset-mount:
        source: oci://skybucket7
        mode: MOUNT

- [v] All smoke tests: `pytest tests/test_smoke.py` 
- [v] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [v] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Review by landscapepainter — COMMENTED** [review]
@HysunHe Awesome! Looking forward to review OCI's Object Storage integration. Gave a quick look and left some comments. Will give a thorough look and test out within this week. Also, here's just a heads up that we are working on a separate class to handle rclone in #1966 so your implementation on `rclone mount` might require an update as we go.

File: sky/data/storage.py [hunk 1]
```diff
@@ -2036,3 +2092,415 @@ def _delete_r2_bucket(self, bucket_name: str) -> bool:
...
+
+            # Check for valid characters and start/end with a number or letter
+            pattern = r'^[A-Za-z0-9-._]+$'
+            if not re.match(pattern, name):
+                _raise_no_traceback_name_error(
+                    f'Invalid store name: name {name} can only contain '
+                    'upper or lower case letters, numeric characters, dashes '
+                    '(-), underscores (_), and dots (.). Spaces are not '
+                    'allowed. Names must start and end with a number or '
+                    'letter.')
```
[the remark this answers is not in the document]
Comment (HysunHe) [reply sky/data/storage.py:2197]: Thanks @landscapepainter for the quick action. For the bucket naming convention, here has some information: https://docs.oracle.com/en-us/iaas/Content/Object/Tasks/managingbuckets.htm

**Comment by HysunHe** [comment]
Thanks @landscapepainter for the quick review. For rclone in this case maybe I use it in a much simpler way, which is just for mounting remote buchets to the provisioned machine. Happy to see some potential improvement as we go...

**Review by landscapepainter — COMMENTED** [review]
@HysunHe Left some comments! I noticed that OCI object storage doesn't have an implementation for `test_smoke.py::TestStorageWithCredentials` yet. Let's add the storage tests as well.

File: sky/data/storage.py [hunk 2]
```diff
@@ -2035,3 +2091,418 @@ def _delete_r2_bucket(self, bucket_name: str) -> bool:
...
+        """
+        try:
+            get_bucket_response = self.client.get_bucket(
+                namespace_name=self.namespace, bucket_name=self.name)
+            bucket = get_bucket_response.data
+            bucket_compartment = bucket.compartment_id
+            if bucket_compartment != self.compartment:
+                raise exceptions.StorageBucketGetError(
+                    f'InvalidBucketLocation: Bucket {self.name} should be '
+                    f'in the same compartment with the sky compute instances.')
```
[the remark this answers is not in the document]
Comment (HysunHe) [reply sky/data/storage.py:2359]: No. The OCI object storge is available regardless of OCI compute instance existence/setup. 

Compute instances provisioned by skypilot will be in skypilot_compartment or root compartment. To make sure buckets can be mounted to these compute instances successfully using rclone, we expect the buckets are in the same compartment as the compute instances.

**Comment by Michaelvll** [comment]
This PR fixes #2072, for GitHub to recognize this.

**Pull request closed without merging**